data_model/src/lib.rs

```diff
@@ -1591,6 +1591,17 @@ pub mod ffi {
     #[cfg(all(feature = "ffi_export", not(feature = "ffi_import")))]
     ffi_fn! {def_ffi_fn}
 
+    #[cfg(any(crate_type = "dylib", crate_type = "cdylib"))]
+    #[cfg(all(feature = "ffi_export", not(feature = "ffi_import")))]
+    mod dylib {
+        #[cfg(not(feature = "std"))]
+        use alloc::alloc;
+        #[cfg(feature = "std")]
+        use std::alloc;
+
+        iroha_ffi::def_ffi_fn! {dealloc}
+    }
```

Comment:
unlike other extern functions, there should be only one `__dealloc` function per generated dynamic library

Reply:
And there's no way of enforcing this invariant at compile time. I doubt the macro trickery is worth it, I'd just add a comment in the text of the program to that effect.

Reply:
I believe this can only be enforced at link time. I will add a comment

This function could be defined once per crate, but there is no need as it would be the same for every crate